File: ios/StrainWise/Strain/NoteBadge.swift
import SwiftUI

/// Small pencil badge that lights up next to a strain whenever the
/// signed-in user has at least one saved note on it. Reads from the
/// shared `SavedStrainsStore` so it stays in sync as notes are added
/// or removed anywhere in the app. Renders nothing for unauthenticated
/// users (the env store is always present, but the slug won't be in its
/// items so `notes(for:)` returns `[]`).
///
/// Mirrors the web `StrainNoteIndicator` (`NotebookPen` icon, primary
/// color, icon-only at the default size). Pair it inline next to any
/// strain name — header hero, compare chip, recommendation card,
/// terpene family row, etc.
struct NoteBadge: View {
    /// Profile or slug the badge belongs to. `StrainProfile` is preferred
    /// because the view recomputes on the slug path; passing a name is
    /// also fine since `slug` falls back to it.
    let profile: StrainProfile

    /// Icon size in points. Default matches the existing
    /// `StrainPoster` indicator (compact / non-compact aware via callers).
    var size: CGFloat = 13

    /// Pass `true` when the badge sits in a tighter container (a chip
    /// or compact card) so the spacing tightens up. Doesn't change the
    /// icon, just the leading padding from the surrounding text.
    var compact = false

    @Environment(SavedStrainsStore.self) private var saved

    private var count: Int {
        saved.notes(for: profile.slug).count
    }

    var body: some View {
        if count > 0 {
            Image(systemName: "square.and.pencil")
                .font(.system(size: size, weight: .semibold))
                .foregroundStyle(Palette.primary)
                .padding(.leading, compact ? 2 : 4)
                .accessibilityHidden(true)
        }
    }
}

extension NoteBadge {
    /// Voiceover-friendly label matching the web `StrainNoteIndicator`
    /// copy. Use as `accessibilityLabel(...)` on the parent view that
    /// also names the strain so VoiceOver hears "Blue Dream, 2 notes".
    static func accessibilityLabel(for profile: StrainProfile, count: Int) -> String {
        guard count > 0 else { return profile.name }
        let noun = count == 1 ? "note" : "notes"
        return "\(profile.name), \(count) \(noun)"
    }
}

File: android/app/src/main/java/ai/strainease/app/ui/components/NoteBadge.kt
package ai.strainease.app.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Edit
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import ai.strainease.app.data.SavedStrainsStore
import ai.strainease.app.models.StrainProfile
import ai.strainease.app.ui.theme.StrainEaseTypography

/**
 * Small "✎" pill that lights up next to a strain whenever the
 * signed-in user has at least one saved note on it. 1:1 port of
 * the iOS `NoteBadge` (Strain/NoteBadge.swift) that mirrors the
 * web `StrainNoteIndicator`.
 *
 * Reads from [SavedStrainsStore.notesFor] which is backed by the
 * `users/{uid}/savedStrains/{slug}` Firestore document's
 * embedded `notes[]` array. Renders nothing when the user has
 * no notes on this strain.
 *
 * Pair it inline next to any strain name (header, compare chip,
 * recommendation card, etc.) so the patient can see at a glance
 * which strains they've journaled.
 */
@Composable
fun NoteBadge(
    profile: StrainProfile,
    savedStrains: SavedStrainsStore,
    modifier: Modifier = Modifier,
    compact: Boolean = false,
) {
    val saved by savedStrains.savedFlow.collectAsState(initial = emptyList())
    val count = remember(saved, profile.slug) {
        saved.firstOrNull { it.slug == profile.slug }?.noteCount ?: 0
    }
    if (count == 0) return
    val padH = if (compact) 6.dp else 8.dp
    val padV = if (compact) 3.dp else 4.dp
    val iconSize = if (compact) 11.dp else 13.dp
    Row(
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = androidx.compose.foundation.layout.Arrangement.spacedBy(2.dp),
        modifier = modifier
            .clip(CircleShape)
            .background(MaterialTheme.colorScheme.primary.copy(alpha = 0.12f))
            .border(1.dp, MaterialTheme.colorScheme.primary.copy(alpha = 0.35f), CircleShape)
            .padding(horizontal = padH, vertical = padV),
    ) {
        Icon(
            imageVector = Icons.Filled.Edit,
            contentDescription = null,
            tint = MaterialTheme.colorScheme.primary,
            modifier = Modifier.size(iconSize),
        )
        if (!compact) {
            Text(
                text = count.toString(),
                style = StrainEaseTypography.labelSmall.copy(fontWeight = FontWeight.SemiBold),
                color = MaterialTheme.colorScheme.primary,
            )
        }
    }
}
